contracts/secretswap_router/src/contract.rs

```diff
@@ -0,0 +1,378 @@
+                }
+                None => Err(StdError::generic_err("cannot find route")),
+            }
+        }
+        HandleMsg::FinalizeRoute {} => match read_route_state(&deps.storage)? {
+            Some(RouteState {
+                is_done,
+                remaining_route,
+            }) => {
+                // this function is called only by the route creation function
+                // it is intended to always make sure that the route was completed successfully
+                // otherwise - revert the transaction
+
+                if env.contract.address != env.message.sender {
+                    return Err(StdError::unauthorized());
+                }
+                if !is_done {
+                    return Err(StdError::generic_err("cannot finalize: route is not done"));
+                }
+                if remaining_route.hops.len() != 0 {
+                    return Err(StdError::generic_err(
+                        "cannot finalize: route still has hops",
+                    ));
+                }
+
+                delete_route_state(&mut deps.storage);
+
+                Ok(HandleResponse::default())
+            }
+            None => Err(StdError::generic_err("no route to finalize")),
+        },
+        HandleMsg::RegisterTokens { tokens } => {
+            let owner = read_owner(&deps.storage)?;
+            if owner != env.message.sender {
+                return Err(StdError::unauthorized());
+            }
+
+            let mut registered_tokens = read_tokens(&deps.storage)?;
+            let mut msgs = vec![];
+
+            for token in tokens {
+                if token.address == None || token.code_hash == None {
+                    return Err(StdError::generic_err(format!(
+                        "token must have an 'address' and a 'code_hash': {:?}",
+                        token
+                    )));
+                }
+
+                let address = token.address.unwrap();
+                let code_hash = token.code_hash.unwrap();
+
+                if registered_tokens.contains(&address) {
+                    continue;
+                }
+                registered_tokens.push(address.clone());
+
+                msgs.push(snip20::register_receive_msg(
+                    env.contract_code_hash.clone(),
+                    None,
+                    256,
+                    code_hash.clone(),
+                    address.clone(),
+                )?);
+                msgs.push(snip20::set_viewing_key_msg(
+                    "SecretSwap Router".into(),
+                    None,
+                    256,
+                    code_hash.clone(),
+                    address.clone(),
+                )?);
+            }
+
+            store_tokens(&mut deps.storage, &registered_tokens)?;
+
+            Ok(HandleResponse {
+                messages: msgs,
+                log: vec![],
+                data: None,
+            })
+        }
+        RecoverFunds {
```

Comment:
Do we want to add a RecoverSCRT as well? (in case there's leftover SCRT, or someone does something stupid and sends SCRT to the contract directly?)

Reply:
Thanks! I forgot about this.